### ISSUE TITLE
sampling: dist_build full-qsorts all 151K vocab entries per sampled token — partial top-p select is ~7x cheaper (measured)

### ISSUE THREAD

**Issue by KingIcyCreamProjects** [JustVugg/colibri#335]
`dist_build()` (`c/glm.c:4193` on dev) sorts the **entire vocabulary** (`qsort(g_pidx, V, ...)` at `:4201`) on every sampled step whenever `0 < NUCLEUS < 1` — which is the serve default (temp resolves to 0.7, NUCLEUS=0.90), so it runs once per emitted token in normal serving, and additionally per draft position under rejection sampling (`:4343`).

Measured on a Ryzen 9950X3D (gcc 16.1, V=151936, GLM-5.2): the qsort with its indirect comparator costs **5.6–8.0 ms/call**; a partial selection (extract descending probabilities until cumulative mass ≥ `g_nuc` — typically a few hundred tokens at p=0.9) costs **~0.78 ms** → **~5 ms/token** of pure serial critical-path CPU saved. On disk-bound configs that's under the noise floor, but it's free tail latency on every config, it compounds with drafting, and it grows into a real fraction as I/O keeps improving.

**The correctness invariant** for any implementation: `dist_sample()` (`:4210`) iterates all V entries of `g_pbuf` directly and relies on `dist_build` having **zeroed the truncated tail** (`:4204`). A partial select must still zero every out-of-head entry, or the sampling distribution silently changes — no crash, just wrong sampling.

Implementation + invariant test on a branch: https://github.com/KingIcyCreamProjects/colibri/tree/perf/topp-partial-select

- max-heap over the index array: heapify O(V), pop descending until cumulative mass reaches `g_nuc`; winners land in `g_pidx[n..V-1]`, the remaining heap prefix is exactly the tail to zero
- `s2` is accumulated in descending order like the old code, so the renormalized head is **bit-identical** (boundary ties were unspecified under the unstable qsort and stay so)
- `tests/test_topp.c` drives the real `dist_build` (the `test_stops.c` include pattern) against an independent reference of the old algorithm: 6 sizes × 4 p values × 5 shapes including all-equal and plateau tie cases plus the `nuc>=1` guard-off path; asserts keep count, bitwise head equality on tie-free shapes, multiset equality on ties, exact-zero tails, and renormalization

Happy to open the PR.

**Comment by woolcoxm** [comment]
working on this now, give me 20 minutes.

**Comment by woolcoxm** [comment]
bug should be fixed will push in 10 minutes currently building tests.

**Comment by woolcoxm** [comment]
**Fix posted:** #354 (`fix/topp-partial-select-335` → `dev`).

## What changed
Replaced the full `qsort(g_pidx, V, ...)` + linear scan in `dist_build()` with a **max-heap partial select**:
- Floyd heapify `g_pidx` over V by descending `g_pbuf` prob — **O(V)**, cache-friendly, no recursion
- Pop winners to the array's high end until `cum >= g_nuc` — `k · O(log V)`
- The remaining heap prefix `g_pidx[0..out-1]` *is* the tail → zero it, renormalize the head

Winners land in descending order, so `s2` accumulates in the same order as before → head is unchanged on tie-free shapes (boundary ties were already unspecified under the unstable qsort). All four `dist_sample` invariants hold (g_pbuf stays id-indexed, g_pidx stays internal, tail fully zeroed, head renormalizes to 1). No API change, no caller change, no new globals.

## Test coverage added
`c/tests/test_topp.c` drives the **real** `dist_build` (via `#include "../glm.c"`) against an independent `double`-precision reimplementation of the **old** algorithm. **123 cases**, all passing locally (MinGW, `-O3 -Wall -Wextra`):
- 6 sizes (1, 2, 8, 64, 257, 1519 ≈ V/100 of GLM-5.2) × 5 shapes (uniform / peaked / geometric / plateau-ties / sharptail) × 4 `g_nuc` values (0.001, 0.5, 0.9, 0.999)
- `g_nuc ≥ 1` and `g_nuc ≤ 0` guard-off paths
- V=1 edge case
- Tie-free shapes → head values within 1e-6 rel; tie shapes (incl. boundary ties) → value-multiset equality

Wired into `TEST_BINS` + a Makefile rule, so it runs in `make test-c` / `make check`. No scratch files, so it builds clean on Windows MinGW without the unmerged `mkdtemp` shim (#352).

## ⚠️ Needs verification before merge
This is the part I can't do alone — the unit test proves **correctness** (output distribution bit-for-bit equivalent to the old algorithm within float noise), but the headline number in this issue is a **latency win**. That needs independent confirmation:

1. **Reproduce the benchmark.** The issue reports 5.6–8.0 ms/call → ~0.78 ms on a Ryzen 9950X3D. Someone on that hardware (or any fast x86) should re-time `dist_build` before/after on a real V=151936 softmax to confirm the ~7× speedup holds and didn't regress under `-O3`. My local box can't measure this meaningfully (the win is structural, not something I verified with a timer).
2. **Serve-decode sanity check.** Run a real `coli serve` decode with `NUCLEUS=0.90` (the default path this changes) for a few hundred tokens and confirm output quality is unchanged — the unit test covers the math, but an end-to-end generation pass catches anything the synthetic shapes miss (e.g. real-logit distributions that don't match any of my 5 shapes).
3. **CI completion.** PR #354's Linux/Python/CUDA-syntax/Web-UI checks are green; **Windows + macOS + CUDA-build were still running at post time** — those need to finish clean. Note: Windows CI will hit the **pre-existing** `test_stops.c`/`mkdtemp` breakage from #352 (independent of this PR); `test_topp` itself builds clean there.

Happy to iterate on any of the above. The benchmark repro (1) is the most important — without it this PR is "provably correct, plausibly faster."

---

> **Credit / co-author:** the max-heap partial-select algorithm and the original 9950X3D ~7× measurement are **@KingIcyCreamProjects**'s work from issue #335 — this PR/branch implements their idea. Attribution is recorded here in-thread rather than as a `Co-Authored-By:` commit trailer only because these commits are already merged into `origin/dev`; amending them for the trailer would rewrite shared integration history (every contributor's `dev` diverges). If proper git-native co-authorship is wanted, that needs a maintainer to file it as a tracked issue/decision on the repo.

**Comment by woolcoxm** [comment]
i will fire up tests so we can get more than 1 system tested to verify the fix, not sure if my processor is the kind its looking for but i will run the tests and post the result.

**Comment by woolcoxm** [comment]
apparently we have some bad algorithms, i will have to look into this further, it was inefficient. i wonder if there are other algorithms we can gain speed from.

will upload my benchmark to the pr for you to run.

2000 iterations × 12 configs × 2 algorithms × warmups on V=151936 — takes a bit

**Comment by woolcoxm** [comment]
**Update — benchmark posted, numbers measured (not just claimed).**

Added `bench_topp` to PR #354 (commit `37c96ee`): a head-to-head microbench that re-implements the **old** full-vocab qsort inline on a private buffer and times it against the **real new** heap partial-select over identical inputs, in one process (controlled A/B — same warm caches, same logits). V=151936, temp=0.7, 2000 timed reps/cell, median reported. `make tests/bench_topp && ./tests/bench_topp`.

Measured on **Intel Core Ultra 9 185H** (16c, Meteor Lake P-cores), `-O3 -march=x86-64-v3`:

```
shape           nuc    old ns/call    new ns/call   speedup      keep
realistic      0.50       22,160,500      10,339,900     2.14x         1
realistic      0.90       23,293,000       5,892,300     3.95x      1370   <- serve default
realistic      0.95       21,946,900       7,499,900     2.93x      3955
realistic      0.99       22,694,400      11,372,500     2.00x     12612
uniform        0.50        1,825,500       1,350,500     1.35x     75969
uniform        0.90        2,058,800       1,608,000     1.28x    136743
uniform        0.95        1,824,200       1,542,300     1.18x    144340
uniform        0.99        1,102,500         929,400     1.19x    150417
plateau        0.50        2,048,300       1,405,200     1.46x        33
plateau        0.90        2,042,900       1,435,700     1.42x        89
plateau        0.95        2,026,800       1,411,400     1.44x       109
plateau        0.99        2,051,800       1,416,800     1.45x       168
```

### Reading the numbers honestly

**The win is real and on the critical path.** At the serve default (`realistic`, `nuc=0.90`), it's **3.95× — ~17 ms/call saved**, and `dist_build` runs once per emitted token plus once per draft position, so that compounds across a decode. The realistic shape (one hot token + exponential tail, i.e. what LM logits actually look like) is where this matters.

**But it is not the 7× from the original report.** Two reasons, both explainable:
1. **Different hardware.** The issue measured 7× on a Ryzen 9950X3D (Zen 5, very wide rename/alu engine, benefits qsort's comparator-heavy loop less than the heap's simpler ones —放大 the ratio). My Core Ultra is slower per-call on *both* sides (old 22 ms vs their 5.6–8 ms; new 5.9 ms vs their 0.78 ms), so the ratio compresses to ~4×.
2. **The win is bounded below by the O(V) pass.** Look at `realistic nuc=0.50 keep=1` — even when only 1 token is kept, new is still 10.3 ms, because heapify touches all 151936 entries. That floor is intrinsic; qsort's O(V log V) is always worse, but the ratio can't exceed (V log V) / V ≈ 18× in the asymptotic best case and is much lower once memory bandwidth dominates.

**Where it helps least:** the `uniform` shape (near-full head, keep≈150K) shows only **1.2×**, because both algorithms do almost all the work and the heap's pop-to-rear overhead nearly cancels its heapify advantage. Real logits are never uniform, so this is a theoretical floor, not a serving regression.

### Net

- **Correctness:** proven by `test_topp` (123 cases, bit-equivalent within float noise).
- **Speed:** reproducibly **2–4× on realistic shapes** at serving nuc values on this hardware; the issue's 7× is plausibly reproducible on a 9950X3D (whoever has one, please re-run `make tests/bench_topp && ./tests/bench_topp` and post your table — the bench is in the PR now).

What I *still* haven't done: an end-to-end `coli serve` decode run to confirm the ms/call savings show up as measurable tok/s improvement. The unit + bench work is done; the serve-level confirmation is the remaining gap.

---

> **Credit / co-author:** the max-heap partial-select algorithm and the original 9950X3D ~7× measurement are **@KingIcyCreamProjects**'s work from issue #335 — this PR/branch implements their idea. Attribution is recorded here in-thread rather than as a `Co-Authored-By:` commit trailer only because these commits are already merged into `origin/dev`; amending them for the trailer would rewrite shared integration history (every contributor's `dev` diverges). If proper git-native co-authorship is wanted, that needs a maintainer to file it as a tracked issue/decision on the repo.

**Comment by woolcoxm** [comment]
doing a deepdive on the algorithms i know of, i bet we can get a few more seconds off stuff.

**Comment by woolcoxm** [comment]
boy am i finding a lot of inefficiencies LOL, lets just say dont use a prompt with a lot of special tokens >>

i never thought to look at the algorithms we are using.

**Comment by woolcoxm** [comment]
after looking there is only one major issue that i can see, and the special token thing, not sure if i can fix that but ill have a look at it.

**Comment by woolcoxm** [comment]
Cross-link: the same partial-select idea, applied to the **attention** side, is now filed as **#356** (branch `fix/dsa-partial-select-356`, based off `dev`).

`attention_rows()` DSA indexer was doing the *exact* antipattern this issue fixed in sampling: full-qsorting an array (`nk` context scores) per layer per token just to read one threshold value. Replaced with `partial_select_desc` (quickselect, `O(nk)` average) + `thr = min(selected)`; the position-order scans are unchanged so the kept set is **bit-identical** — a stronger contract than the sampling heap here could promise (multiset-only, because the heap was unstable).

Measured 10–21× on the same style of standalone bench (median of 2000 reps), and the win grows with context length (linear vs n-log-n). Ships with `test_dsa_select` (129 cases) + `bench_dsa_select`, mirroring the `test_topp`/`bench_topp` pair from #354.

Also recorded in #356: three other attention-side candidates I measured and deliberately did **not** file (MoE top-K, softmax, tok_encode) — all already optimal at GLM-5.2 scale.

---

> **Credit / co-author:** the max-heap partial-select algorithm and the original 9950X3D ~7× measurement are **@KingIcyCreamProjects**'s work from issue #335 — this PR/branch implements their idea. Attribution is recorded here in-thread rather than as a `Co-Authored-By:` commit trailer only because these commits are already merged into `origin/dev`; amending them for the trailer would rewrite shared integration history (every contributor's `dev` diverges). If proper git-native co-authorship is wanted, that needs a maintainer to file it as a tracked issue/decision on the repo.

**Comment by KingIcyCreamProjects** [comment]
Shipped ✅ — #354 (heap partial-select) merged to `dev`. Independent 9950X3D repro is on the PR: **13.9× at the serve default (nuc=0.90)**, up to 17.4×, landing at 7.08× where keep is largest — https://github.com/JustVugg/colibri/pull/354#issuecomment-5006784532. Correctness locked by `test_topp` (123 cases, bit-equivalent).

This will auto-close when `dev` → `main`. Only non-blocking follow-up left is the serve-decode tok/s sanity check — I can run that on the box if it's useful, otherwise this is done. Thanks @woolcoxm for the fast turnaround.

**Comment by JustVugg** [comment]
Fixed on `dev` — **#354** is merged. `dist_build` no longer full-qsorts the 151,936-entry vocab; it uses a max-heap partial select that finds only the top-p head. Your measured ~7× is the motivation, and the merge carries an equivalence test (`tests/test_topp.c`, 123 cases) that asserts the renormalized nucleus is **bit-identical** to the old full-qsort path on tie-free shapes — so it's faster with provably-unchanged sampling. Thanks for the measurement and the clean writeup. Closing.
